modules/3d/src/rgbd/odometry.cpp

```diff
@@ -2,2475 +2,343 @@
 // of this distribution and at http://opencv.org/license.html
 
-// FastICPOdometry is based on kinfu-remake code
-// Copyright(c) 2012, Anatoly Baksheev
-// All rights reserved.
-
 #include "../precomp.hpp"
 #include "utils.hpp"
-#include "fast_icp.hpp"
+#include "opencv2/3d/odometry.hpp"
+#include "odometry_functions.hpp"
 
 namespace cv
 {
 
-enum
+class Odometry::Impl
 {
-    RGBD_ODOMETRY = 1,
-    ICP_ODOMETRY = 2,
-    MERGED_ODOMETRY = RGBD_ODOMETRY + ICP_ODOMETRY
+private:
+
+public:
+    Impl() {};
+    virtual ~Impl() {};
+    virtual OdometryFrame createOdometryFrame() = 0;
+    virtual void prepareFrame(OdometryFrame frame) = 0;
+    virtual void prepareFrames(OdometryFrame srcFrame, OdometryFrame dstFrame) = 0;
+    virtual bool compute(OdometryFrame srcFrame, OdometryFrame dstFrame, OutputArray Rt) const = 0;
 };
 
-static const int sobelSize = 3;
-static const double sobelScale = 1./8.;
-static const int normalWinSize = 5;
-static const RgbdNormals::RgbdNormalsMethod normalMethod = RgbdNormals::RGBD_NORMALS_METHOD_FALS;
 
-static
-void buildPyramidCameraMatrix(const Matx33f& cameraMatrix, int levels, std::vector<Matx33f>& pyramidCameraMatrix)
+class OdometryICP : public Odometry::Impl
 {
-    pyramidCameraMatrix.resize(levels);
+private:
+    OdometrySettings settings;
+    OdometryAlgoType algtype;
 
-    for(int i = 0; i < levels; i++)
-    {
-        Matx33f levelCameraMatrix = (i == 0) ? cameraMatrix : 0.5f * pyramidCameraMatrix[i-1];
-        levelCameraMatrix(2, 2) = 1.0;
-        pyramidCameraMatrix[i] = levelCameraMatrix;
-    }
-}
+public:
+    OdometryICP(OdometrySettings settings, OdometryAlgoType algtype);
+    ~OdometryICP();
+
+    virtual OdometryFrame createOdometryFrame() override;
+    virtual void prepareFrame(OdometryFrame frame) override;
+    virtual void prepareFrames(OdometryFrame srcFrame, OdometryFrame dstFrame) override;
+    virtual bool compute(OdometryFrame srcFrame, OdometryFrame dstFrame, OutputArray Rt) const override;
+};
 
-static inline
-void checkImage(InputArray image)
+OdometryICP::OdometryICP(OdometrySettings _settings, OdometryAlgoType _algtype)
 {
-    if(image.empty())
-        CV_Error(Error::StsBadSize, "Image is empty.");
-    if(image.type() != CV_8UC1)
-        CV_Error(Error::StsBadSize, "Image type has to be CV_8UC1.");
+    this->settings = _settings;
+    this->algtype = _algtype;
 }
 
-static inline
-void checkDepth(InputArray depth, const Size& imageSize)
+OdometryICP::~OdometryICP()
 {
-    if(depth.empty())
-        CV_Error(Error::StsBadSize, "Depth is empty.");
-    if(depth.size() != imageSize)
-        CV_Error(Error::StsBadSize, "Depth has to have the size equal to the image size.");
-    if(depth.type() != CV_32FC1)
-        CV_Error(Error::StsBadSize, "Depth type has to be CV_32FC1.");
 }
 
-static inline
-void checkMask(InputArray mask, const Size& imageSize)
+OdometryFrame OdometryICP::createOdometryFrame()
```

Comment:
By the way:
Sometimes a user should call `setPyramidLevel` to resize pyramid vector, see any kinfu pipeline code.
Since the `Odometry` object is already created and it knows what the vector size should be -
can we do this resize automatically, at `createOdometryFrame()` call?

Reply:
no, only after creating